_posts/2023-11-17-transactional-outbox.md

```diff
@@ -0,0 +1,295 @@
+Another significant advantage of the Transactional Outbox Pattern is the decoupling it offers. By separating the concerns of database transactions and event publishing, the pattern allows each to operate independently. This separation not only simplifies the architecture but also enhances scalability and maintainability. 
+It allows the system to evolve more flexibly, adapting to changes and scaling components independently without disrupting the fundamental operations of transaction management and event distribution.
+
+Additionally, an often overlooked yet significant aspect of the Transactional Outbox Pattern is its efficiency, 
+particularly regarding the speed of writing to the outbox. 
+Writing an event to the outbox involves merely appending a row to a table, which is one of the fastest operations in database management. This efficiency means that the pattern does not impose a notable performance penalty.
+
+# Challenges and Solutions
+
+Implementing the Transactional Outbox Pattern, while beneficial, brings its own set of challenges that need careful consideration and handling.
+
+## At least once delivery
+
+When implementing the Transactional Outbox Pattern in a distributed system, it's important to understand that this pattern inherently guarantees at-least-once-delivery semantics. 
+At-least-once-delivery is a delivery approach where the system ensures that a message is definitely delivered to its destination at least once.
+Since the delivery mechanism is designed to ensure that messages are not lost, it errs on the side of sending duplicates rather than risking a message not being delivered at all. 
+As a result, the receiving services or components must be equipped to identify and appropriately handle these duplicate messages. 
+This involves implementing logic to detect if a message has already been processed and ensuring that processing it again does not adversely affect the system's state or lead to incorrect operations.
+
+A common approach to handle duplicate messages is implementing idempotency at the technical level. 
+This involves recording the IDs of consumed events and checking against this record before processing new events. 
+If an incoming event's ID matches one that has already been processed, it is ignored, thus preventing duplicate processing.
+
+A more sophisticated approach is to design idempotency based on business rules. 
+Instead of just tracking event IDs, this approach involves designing the business logic and data models in such a way that repeating the same operation does not have adverse effects. 
+For example, applying the same update twice would result in the same state as applying it once. This method requires a deeper understanding of the business context but can lead to a more robust solution.
```

Comment:
I think this is clear, but some people might like an example? like: instead of each event recording the number of tacos I ate in each meal, record how many are left after each meal. this way, we can just pick the smallest number. seeing the same event twice has no effect. when recording how many tacos I ate, seeing the same event twice might result in a negative number of tacos (which is definitely a catastrophe for me)